### PR TITLE
nbhood: preserve dimensionality and units where possible

### PR DESCRIPTION
**Pull request by benfitzpatrick**
This change avoids introducing extra realization or time dimensions during the neighbourhood processing process.

This happens particularly for the lead time varying radii change since it has to slice over time. For a single cube with scalar time, an additional time dimension would have been added. This is different behaviour from the fixed radius path.

An exception is the occurrence within vicinity, since that explicitly collapses realization.

This change also fixes an accidental permanent conversion of the forecast_period coordinate into hours during the lead time calculation.

Updated existing unit tests - result cubes should always match input cube dimensionality now (except for occurrence within vicinity realization collapse).

### CODE REVIEW

File: lib/improver/nbhood/nbhood.py [hunk 1]
```diff
@@ -203,7 +203,7 @@ def process(self, cube, mask_cube=None):
         # the cube is sliced, so that the realization becomes a scalar
         # coordinate.
         try:
-            realiz_coord = cube.coord('realization')
+            realiz_coord = cube.coord('realization', dim_coords=True)
```
Comment (wxtim) [inline lib/improver/nbhood/nbhood.py]: Additionally checks that realization is a dim-coord.
@benfitzpatrick: Should this change be reflected in the comment above?

File: lib/improver/nbhood/nbhood.py [hunk 2]
```diff
@@ -221,7 +221,7 @@ def process(self, cube, mask_cube=None):
         if np.isnan(cube.data).any():
             raise ValueError("Error: NaN detected in input cube data")
 
-        cubelist = iris.cube.CubeList([])
+        cubelist = []
```
Comment (wxtim) [inline lib/improver/nbhood/nbhood.py]: is it confusing to call a list of cubes a cubelist?

**Comment by wxtim** [comment]
I'd like to see a unit test that shows how the existing code breaks, and therefore how the new code fixes that breakage.

**Comment by benfitzpatrick** [comment]
Hi Tim,

There were already tests capturing this but they captured the old non-desirable behaviour! The best ones are the existing ones I've had to modify that are named `without_realization` - they now don't suddenly gain a dimensional realization coordinate! I'll try and address your other points, and the ones from the bot!

**Comment by wxtim** [comment]
Yes, I see now. I wonder if this should be tested more explicitly - something like
```
self.assertEqual(cube[0].shape, result.shape)
```

**Comment by benfitzpatrick** [comment]
That would be good! OK

File: lib/improver/nbhood/nbhood.py (hunk 1 above)
Comment (benfitzpatrick) [reply lib/improver/nbhood/nbhood.py]: Basically it is a double check that it is not scalar

Comment (benfitzpatrick) [reply lib/improver/nbhood/nbhood.py]: Updated comment

File: lib/improver/nbhood/nbhood.py (hunk 2 above)
Comment (benfitzpatrick) [reply lib/improver/nbhood/nbhood.py]: Updated

**Comment by benfitzpatrick** [comment]
> I'd like to see a unit test that shows how the existing code breaks, and therefore how the new code fixes that breakage.
adding
>
> self.assertEqual(result.coord('forecast_period').units,
>                         Unit('s'))
> to test_radii_varying_with_lead_time would check that the output units are correct.

Unit unit test added!

**Comment by benfitzpatrick** [comment]
New test in 137eeaa, new updates to tests and addressing of review comments in e9cebe1

File: lib/improver/blending/weighted_blend.py [hunk 3]
```diff
@@ -113,6 +113,8 @@ def conform_metadata(
                 forecast_period_coord(cube,
                                       force_lead_time_calculation=True))
             forecast_period.bounds = None
+            forecast_period.convert_units(cube.coord("forecast_period").units)
```
Comment (bayliffe) [inline lib/improver/blending/weighted_blend.py]: Pretty sure forecast period should always be in seconds, which is what the forecast_period_coord plugin above returns. Does it need to match forecast_period for some reason?

File: lib/improver/blending/weighted_blend.py [hunk 4]
```diff
@@ -113,6 +113,8 @@ def conform_metadata(
...
+            forecast_period.var_name = cube.coord("forecast_period").var_name
```
Comment (bayliffe) [inline lib/improver/blending/weighted_blend.py]: I guess this fixed the bug I was seeing when running the suite, so hooray.

File: lib/improver/blending/weighted_blend.py (hunk 3 above)
Comment (benfitzpatrick) [reply lib/improver/blending/weighted_blend.py]: This preserves the behaviour in our tests - some of those have forecast_period in hours.

File: lib/improver/blending/weighted_blend.py (hunk 4 above)
Comment (benfitzpatrick) [reply lib/improver/blending/weighted_blend.py]: Yep!

**Comment by bayliffe** [comment]
Thanks for the fix Gavin. I have had a look, run the tests and updated the acceptance tests after checking the only difference is the demotion of what were new dimension coordinates back to scalars.

**Pull request merged**